### PR TITLE
feat: add preview search to new API

### PR DESCRIPTION
**Pull request by ybt195**
## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: master/internal/api_experiment.go [hunk 1]
```diff
@@ -43,3 +53,55 @@ func (a *apiServer) GetExperiments(
...
+
+	s := searcher.NewSearcher(
+		req.Seed, searcher.NewSearchMethod(config.Searcher), config.Hyperparameters)
+	sim, err := searcher.Simulate(s, nil, searcher.RandomValidation, true, config.Searcher.Metric)
+	if err != nil {
+		return nil, err
+	}
+	protoSim := &experimentv1.ExperimentSimulation{Seed: req.Seed}
+	indexes := make(map[string]int)
+	toProto := func(k searcher.Kind) experimentv1.WorkloadKind {
```
Comment (brain-good) [inline master/internal/api_experiment.go:79]: question: do we think that this func may be useful outside of the context of `PreviewSearch`? I feel like that would be reasonable and we'd want to take this function out of `PreviewSearch`.

File: master/internal/api_experiment.go [hunk 2]
```diff
@@ -43,3 +53,55 @@ func (a *apiServer) GetExperiments(
...
+			return experimentv1.WorkloadKind_WORKLOAD_KIND_UNSPECIFIED
+		}
+	}
+	for _, result := range sim.Results {
+		var workloads []experimentv1.WorkloadKind
+		for _, msg := range result {
+			w := toProto(msg.Workload.Kind)
+			workloads = append(workloads, w)
+		}
+		hash := fmt.Sprint(workloads)
```
Comment (brain-good) [inline master/internal/api_experiment.go:97]: honestly this got me a little lost:

1) it's not a hash technically
2) I'm not sure what we're trying to do here

File: proto/src/determined/experiment/v1/searcher.proto [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+option go_package = "github.com/determined-ai/determined/proto/pkg/experimentv1";
+
+import "google/protobuf/struct.proto";
+
+// WorkloadKind defines the kind of workload that should be executed by trial runners.
+enum WorkloadKind {
+    // Denotes an unknown workload kind.
+    WORKLOAD_KIND_UNSPECIFIED = 0;
+    // Signals to a trial runner that it should run a training step.
+    WORKLOAD_KIND_RUN_STEP = 1;
```
[the remark this answers is not in the document]
Comment (ybt195) [reply proto/src/determined/experiment/v1/searcher.proto:13]: No, this is just a copy of the enum from before. We can address this at another time.

File: proto/src/determined/api/v1/api.proto [hunk 4]
```diff
@@ -63,6 +63,8 @@ service Determined {
 
     // Get a list of experiments.
     rpc GetExperiments(GetExperimentsRequest) returns (GetExperimentsResponse) { option (google.api.http) = {get: "/api/v1/experiments"}; }
+    // Preview hyperparameter search.
+    rpc PreviewSearch(PreviewSearchRequest) returns (PreviewSearchResponse) { option (google.api.http) = {post: "/api/v1/searchers/preview" body: "*"}; }
```
[the remark this answers is not in the document]
Comment (ybt195) [reply proto/src/determined/api/v1/api.proto:67]: Recommendation?

File: master/internal/api_experiment.go (hunk 1 above)
Comment (ybt195) [reply master/internal/api_experiment.go:79]: Not at the moment, if we do, then we can pull it out then.

File: proto/src/determined/api/v1/api.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (brain-good) [reply proto/src/determined/api/v1/api.proto:67]: `hp-searchers`?

**Pull request merged**